### DISCUSSION TITLE
Apply a swirl effect to the background

### DISCUSSION THREAD

**Discussion by Lzrs111** [excaliburjs/Excalibur#2684] [Q&A] [answered]
Hi,

I'm looking for a way to apply a swirl effect to some random part of my background, but the issue is I need to remove anything else in the foreground before doing so. Currently I have found a way to remove stuff by doing graphics.visible = false for a single frame and taking a screenshot, but this is a bit ugly since it's noticeable. Is there any way to access the draw function and do it there after the background, but before everything else gets drawn?

Second, let's say I have my screenshot and I store it in some variable. How would I pass this variable to the Canvas class to be used in its draw function?

Thanks.

**Comment by eonarheim** [comment]
Hi @Lzrs111 

This is an interesting question! Can you provide an example image that's close to what you're trying to achieve?

**Reply by Lzrs111** [reply]
Before:
![Blue Nebula 8 - 512x512](https://github.com/excaliburjs/Excalibur/assets/19205191/07c8f783-44bc-4025-a645-c791f6a6ceed)

After:
![vortex](https://github.com/excaliburjs/Excalibur/assets/19205191/184a6002-f258-4194-b5ab-1f7bf4f7562a)



Something like this :) I think this should be doable in canvas, my issue is that there's enemies or projectiles in the foreground and if I cheat and attach a picture like i just did, it looks weird since I'd like to also animate the effect, not to mention it has to be pixel perfect. I'm going for a sort of black hole/vortex thing. 

So I'm asking if it's possible to:

1. determine where black hole will spawn, say x:643, y: 587
2. remove all foreground actors, including UI in a certain radius from that center point
3. cut out that part of background and add an effect like above to it using the Canvas class
4. restore all foreground actors
5. all that in a single frame

Thanks :)

**Reply by eonarheim** [reply, marked as answer]
Hi @Lzrs111! I think I have something for you

If you're okay trying some alpha code, I've been working on a way to add custom shaders to actors. https://github.com/excaliburjs/Excalibur/pull/2670 `excalibur@0.28.0-alpha.698`

I found a swirl shader example on shader toy https://www.shadertoy.com/view/Xscyzn

Slightly tweaked to use timing:

```typescript
var game = new ex.Engine({
  canvasElementId: 'game',
  width: 512,
  height: 512,
  displayMode: ex.DisplayMode.FitScreenAndFill,
  backgroundColor: ex.Color.Black,
  antialiasing: true
});

var tex = new ex.ImageSource('https://cdn.rawgit.com/excaliburjs/Excalibur/7dd48128/assets/sword.png', false, ex.ImageFiltering.Pixel);
var background = new ex.ImageSource('./stars.png', false, ex.ImageFiltering.Blended);

var loader = new ex.Loader([tex, background]);

var fragmentSource = `#version 300 es
precision mediump float;

// UV coord
in vec2 v_uv;

uniform sampler2D u_graphic;

uniform vec2 u_resolution;

uniform float iTime;

uniform vec2 iMouse;

uniform vec2 u_size;

uniform vec4 u_color;

uniform float u_opacity;

out vec4 fragColor;

void main() {
   vec4 color = u_color;

  float effectRadius = .5;
  float effectAngle = mod(iTime/2., 2.)  * 3.14159;

  vec2 size = u_size.xy;

  vec2 center = iMouse.xy / u_size.xy;
  vec2 uv = v_uv.xy - center;

  float len = length(uv * vec2(size.x / size.y, 1.));
  float angle = atan(uv.y, uv.x) + effectAngle * smoothstep(effectRadius, 0., len);
  float radius = length(uv);
  vec2 newUv = vec2(radius * cos(angle), radius * sin(angle)) + center;

   color = texture(u_graphic, newUv);
   color.rgb = color.rgb * u_opacity;
   color.a = color.a * u_opacity;
  
   fragColor = color * u_color;
}
`

var swirlMaterial = game.graphicsContext.createMaterial({
  name: 'swirl',
  fragmentSource
});

var click = ex.vec(0, 0);

game.input.pointers.primary.on('down', evt => {
  click = evt.worldPos; // might need to change if you have a camera
});

var actor = new ex.Actor({x: 100, y: 100, width: 50, height: 50});
actor.onInitialize = () => {
  var sprite = new ex.Sprite({
    image: tex,
    destSize: {
      width: 300,
      height: 300
    }
  });
  actor.graphics.add(sprite);
};


var backgroundActor = new ex.ScreenElement({x: 0, y: 0, width: 512, height: 512, z: -1});
var time = 0;
backgroundActor.onPostUpdate = (_, delta) => {
  time += (delta / 1000);
  swirlMaterial.getShader().use();
  swirlMaterial.getShader().trySetUniformFloat('iTime', time);
  swirlMaterial.getShader().trySetUniformFloatVector('iMouse', click);
}
backgroundActor.onInitialize = () => {
  backgroundActor.graphics.add(background.toSprite());
  backgroundActor.graphics.material = swirlMaterial;
};

game.add(actor);
game.add(backgroundActor);
game.start(loader);
```

This will put a swirl on the background wherever I click!

<img width="476" alt="image" src="https://github.com/excaliburjs/Excalibur/assets/612071/62be0c5a-1aa5-42ed-9505-7950b474a90b">

## Otherwise postprocessors

Otherwise you can do something similar with postprocessors and shaders but it will apply to the entire screen 

https://excaliburjs.com/docs/postprocessors/#custom-shader-based-post-processors

**Reply by eonarheim** [reply]
I uploaded an example into the test directory https://github.com/excaliburjs/Excalibur/tree/main/sandbox/tests/material

**Comment by Lzrs111** [comment]
Hi @eonarheim ,

Thank you :) I'll play around with it and see how it goes!
